kubernetes/callback.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+
+import logging
+import os
+import socket
+import sys
+import time
+
+from kubernetes import client as k8s_client, config as k8s_config
+from urllib3.exceptions import HTTPError
+from six.moves.http_client import HTTPException
+
+logger = logging.getLogger(__name__)
+
+
+class CoreV1Api(k8s_client.CoreV1Api):
+
+    def retry(func):
+        def wrapped(*args, **kwargs):
+            count = 0
+            while True:
+                try:
+                    return func(*args, **kwargs)
+                except (HTTPException, HTTPError, socket.error, socket.timeout):
+                    if count >= 10:
+                        raise
+                    logger.info('Throttling AWS API requests...')
+                    time.sleep(2 ** count * 0.5)
+                    count += 1
+        return wrapped
+
+    @retry
+    def patch_namespaced_endpoints(self, *args, **kwargs):
+        return super(CoreV1Api, self).patch_namespaced_endpoints(*args, **kwargs)
+
+    @retry
+    def patch_namespaced_pod(self, *args, **kwargs):
+        return super(CoreV1Api, self).patch_namespaced_pod(*args, **kwargs)
+
+
+def patch_master_endpoint(api, namespace, cluster):
+    addresses = [k8s_client.V1EndpointAddress(ip=os.environ['POD_IP'])]
+    ports = [k8s_client.V1EndpointPort(port=5432)]
+    subsets = [k8s_client.V1EndpointSubset(addresses=addresses, ports=ports)]
+    body = k8s_client.V1Endpoints(subsets=subsets)
+    return api.patch_namespaced_endpoints(cluster, namespace, body)
+
+
+def patch_pod_labels(api, namespace, new_role):
+    pod_name = os.environ['POD_NAME']
+    metadata = k8s_client.V1ObjectMeta(namespace=namespace, name=pod_name, labels={'role': new_role})
+    body = k8s_client.V1Pod(metadata=metadata)
+    return api.patch_namespaced_pod(pod_name, namespace, body)
+
+
+def main():
+    logging.basicConfig(format='%(asctime)s %(levelname)s: %(message)s', level=logging.INFO)
+    if len(sys.argv) != 4 or sys.argv[1] not in ('on_start', 'on_stop', 'on_role_change'):
+        sys.exit('Usage: %s <action> <role> <cluster_name>', sys.argv[0])
+
+    action, role, cluster = sys.argv[1:4]
+
+    k8s_config.load_incluster_config()
+    k8s_api = CoreV1Api()
+
+    namespace = os.environ['KUBERNETES_NAMESPACE']
+
+    if role == 'master' and action in ('on_start', 'on_role_change'):
+        patch_master_endpoint(k8s_api, namespace, cluster)
```

Comment:
Doing this as a callback is fragile, If the API call fails the cluster is effectively unavailable until failover is externally triggered. 

Also, there is no interlock to prevent racy updates. Imaginary scenario:

0. unhealthy cluster
1. n-0 acquires leader lock, starts callback.py
2. n-0 hangs for a period of time (IO system problems, network issues, etc.)
3. n-1 acquires leader lock, starts callback.py
4. n-1 callback.py completes, directing endpoint traffic to n-1
5. n-0 becomes active again, runs the API call and overwrites the endpoint with n-0
6. n-0 demotes itself
7. until next failover traffic is now routed to the standby.

I think this needs integration with the HA loop to fix properly.

Reply:
That was just a fast way to copy a script we already have in Spilo: https://github.com/zalando/spilo/blob/master/postgres-appliance/callback_role.py

Agree, it probably should be done from HA loop.

Reply:
There's always going to be the possibility of code halting between checking role='master' and patch_master_endpoint, no?  Or are you suggesting that the master should have a loop which constantly checks that it still owns the endpoint, so that incoherency can be limited to the loop time?

Reply:
I suppose the demotion code needs to have an endpoint check as well.  Although what would a demoting node do if it found it owned the endpoint?

Reply:
Yes I think the loop should check that it still own the endpoint. Also, I think it should release the leader key if it fails to get the endpoint within a reasonable amount of time. I think this might have some overlap with my todo item of implementing basic OCF resource support. Similar considerations would apply with moving a virtual IP around, etc.

Reply:
In order not to have a race conditions master endpoint should be changed only by leader.
Current implementation is using ConfigMaps to store information about leader. The same way we can use Endpoints instead. I.e. Patroni can periodically (every HA loop) update annotations and **subsets** of the leader Endpoint.

Off course such solution has it's own pros and cons.

Pros: when the new master get a leader lock it will atomically change master service

Cons:
 * I have no idea how kubernetes will react on periodic updates of leader `Endpoint`. Every update in theory should cause update of leader `Service`. We should really check that it will not trigger rebuild of iptables rules everywhere.
 * Right now Patroni is does 2 API reads (list_namespaced_config_map and list_namespaced_pod). New implementation will do 3 calls (list_namespaced_endpoint). We kind of can avoid it by converting all ConfigMaps to Endpoints, but I am not sure that it is a good idea.

Reply:
Directly patching Endpoint has another downside - OpenShift does not let normal users make endpoints that point to pods. As documented [in the final paragraph here](https://docs.openshift.com/container-platform/3.5/architecture/core_concepts/pods_and_services.html). At least for OpenShift I would like to rely on pod annotations and have the platform take care of routing a Service to the master. But even that still has the same issue that for reliability a callback is less than optimal.

Reply:
I'm nervous about the idea of dropping the leader lock if grabbing the endpoint/service fails.  That ups the number of sequential operations which need to succeed to three; that seems prone to race-condition type failures.  

Some unresolved questions:

1. What should be the proper ordering of leader lock, failover/promotion, endpoint?  
2. What's the sequence of undoing these actions if any of them fails?  What happens if the undo fails?
3. What do we want to be the result if  kubernetes is unresponsive/misconfigured and no node can update the endpoint?

That last case is potentially pathological, and argues that updating the endpoint needs to come *before* failover/promotion instead of afterwards, since changing the endpoint is easily reversable and failover is not.  If we keep it afterwards, having RBAC config issues or container networking issues could mean that you have a set of disconnected nodes which can't be ever reconnected as replicas, and among which Patroni can't choose a good new master candidate.

I think this needs a deeper discussion than a line-item comment on a patch review.

Reply:
@ants, @jberkus:

I this this commit https://github.com/zalando/patroni/pull/500/commits/baaf20b8cf381bb04773d2b3bd45a528a6d59047 should fix most of the problems.
Since Patroni is already doing some metadata manipulations on the Pod, why not assign label with the `role` right from HA loop? It makes it possible to use labelSelector without any external callback script.

Unfortunately it is still not solving a biggest problem, when the master Pod hangs for a long time and looses the leader key. In this case it still could happen that two Pods will get role=master...

It looks like the only reliable way to make sure that master Endpoint has only one IP is patching it from Patroni HA loop together (and atomically) with updating leader lock.

Reply:
Let's discuss this on the race conditions issue